server/bin/pbench-cull-unpacked-tarballs.py

```diff
@@ -0,0 +1,362 @@
+        # subject objects.
+        for subdir_n in subdir_l:
+            full_p = os.path.join(dir_n, subdir_n)
+            try:
+                link = os.readlink(full_p)
+            except OSError:
+                continue
+            if os.path.realpath(link) == tb_incoming_dir:
+                # FIXME: Remember the path to the removed symbolic link, and
+                # remove any parent directories that become empty as a result
+                # of the symbolic link removal.
+                if dry_run:
+                    print(f"rm {full_p}")
+                else:
+                    try:
+                        os.unlink(full_p)
+                    except OSError as exc:
+                        logger.error("Failed to remove symlink '{}': {}", full_p, exc)
+                        errors += 1
+    return errors
+
+
+def remove_unpacked(tb_incoming_dir, controller_name, results, users, logger, dry_run):
+    """remove_unpacked - Remove the unpacked tar ball directory from the
+    INCOMING tree and all symbolic links to that directory from the RESULTS
+    and USERS trees.
+
+    The symbolic links in the RESULTS and USERS (if any) trees are removed
+    first, then the directory in the INCOMING tree is removed.
+
+    The `tb_incoming_dir` should be a fully resolved path including the tar
+    ball directory name, allowing us to resolve the symbolic link being
+    considered to compare for equality.
+
+    Any errors encountered while removing items will result in those errors
+    being logged.  All removals stop on the first error encountered, so the
+    removal process could leave the unpacked tar ball in a partially deleted
+    state.
+
+    Returns 0 on success, > 0 if any errors encountered.
+    """
+    # Walk the results hierarchy for the tar ball's controller to find all
+    # symbolic links to the INCOMING directory location.  Once we find a
+    # symbolic link (which will be returned in the list of sub-directories)
+    errors = remove_symlinks(
+        os.path.join(results, controller_name), tb_incoming_dir, logger, dry_run
+    )
+    if errors > 0:
+        return errors
+
+    # Fetch user from metadata.log file.
+    user_name = fetch_username(tb_incoming_dir)
+    if user_name:
+        errors = remove_symlinks(
+            os.path.join(users, user_name, controller_name),
+            tb_incoming_dir,
+            logger,
+            dry_run,
+        )
+        if errors > 0:
+            return errors
+
+    # Now that all RESULTS symbolic links, and any USERS symbolic links to the
+    # incoming directory have been removed, we can rename the directory (so it
+    # is no longer visible to the web server) and then delete it.
+    if dry_run:
+        print(f"rm -r {tb_incoming_dir}")
+    else:
+        deleten = f".delete.{os.path.basename(tb_incoming_dir)}"
+        dirn = os.path.dirname(tb_incoming_dir)
+        del_path = os.path.join(dirn, deleten)
```

Comment:
consider use to try: finally:

Reply:
Not sure how `try/finally` would apply here.  If the `os.rename` fails, there is no action that I can see that we always want performed in the finally, same is true for the `shutil.rmtree` call.

Can you given an example of what you mean?